### ISSUE TITLE
Rollup causing hard crash during GC

### ISSUE THREAD

**Issue by jo-sm** [rollup/rollup#1225]
I'm using Rollup to build my application (which includes React and is rolled up into the bundle). Intermittently I get a hard crash during the build phase, related to the GC, which causes the entire watch script to quit to the terminal. Here is an example stack trace with GC data:

```
<--- Last few GCs --->

 6706976 ms: Mark-sweep 1396.4 (1457.8) -> 1396.2 (1457.8) MB, 1282.4 / 0 ms [allocation failure] [GC in old space requested].
 6708281 ms: Mark-sweep 1396.2 (1457.8) -> 1396.2 (1457.8) MB, 1304.9 / 0 ms [allocation failure] [GC in old space requested].
 6709546 ms: Mark-sweep 1396.2 (1457.8) -> 1390.7 (1457.8) MB, 1265.2 / 0 ms [last resort gc].
 6710825 ms: Mark-sweep 1390.7 (1457.8) -> 1392.7 (1457.8) MB, 1278.8 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1a68767b4629 <JS Object>
    1: new constructor(aka DeclaratorProxy) [/node_modules/rollup/dist/rollup.js:~7259] [pc=0x18057c57d477] (this=0x13c551030111 <a DeclaratorProxy with map 0x14f6eebe6e51>,name=0x3f7f34ee78a9 <String[7]: _typeof>,declarator=0x5b5a7ee00b1 <an Object with map 0x3dc6522f9b71>,isTopLevel=0x1a6876704211 <true>,init=0x5b5a7ee0121 <an Object with map 0xcb82f6af249>)
    3: /* ano...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
./start.sh: line 64:   862 Abort trap: 6
```

I'm on 0.39.2 (I thought it might have been caused by saving a JSON version of the cache in case there are syntax errors (this bug was fixed in 0.39.2), but it still occurs after that).

**Comment by kzc** [comment]
Only during rollup --watch?

**Comment by jo-sm** [comment]
I don't use rollup watch, I use chokidar.

**Comment by jo-sm** [comment]
Happened again: 

```
<--- Last few GCs --->

 2356612 ms: Mark-sweep 1380.8 (1458.0) -> 1383.9 (1458.0) MB, 1258.2 / 0 ms [allocation failure] [GC in old space requested].
 2357890 ms: Mark-sweep 1383.9 (1458.0) -> 1383.9 (1458.0) MB, 1277.4 / 0 ms [allocation failure] [GC in old space requested].
 2359173 ms: Mark-sweep 1383.9 (1458.0) -> 1378.8 (1458.0) MB, 1283.5 / 0 ms [last resort gc].
 2360845 ms: Mark-sweep 1378.8 (1458.0) -> 1380.4 (1458.0) MB, 1671.7 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x267488fb4629 <JS Object>
    1: defer [native promise.js:~141] [pc=0xafe73ac76a4] (this=0x267488f9cd49 <JS Function Promise (SharedFunctionInfo 0x267488f677a1)>)
    2: chain [native promise.js:~175] [pc=0xafe735ae783] (this=0x33b01c267e19 <a Promise with map 0x2a032d0f8d59>,t=0x33b01c031289 <JS Function resolve (SharedFunctionInfo 0x28e39dcea8d9)>,u=0x33b01c0312d1 <JS Function reject (SharedFunctionInfo 0x28e39dcea981...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6
```

**Comment by kzc** [comment]
Hard to know how the third party chokidar module is managing its memory. I think it's beyond the scope of rollup.

You might try spawning each watch action within its own node vm instance to prevent memory leaks:

https://nodejs.org/api/vm.html

**Comment by jo-sm** [comment]
I'm pretty sure it's not caused by chokidar though, since the GC doesn't occur on any modules that chokidar specifically uses. Chokidar doesn't spawn on each code change either; it watches once at the start of the script and I only have two watchers set up, and the watchers are lightweight since they use `fs.watchFile` internally. 

My codebase is pretty small, less than 50 components and modules, so my guess is that it's something I'm importing like React and that there's a memory issue, like that something is too large or cumbersome to handle via GC within the memory on my machine, but I don't know how to test for that.

**Comment by kzc** [comment]
Seems that rollup-watch as of 2 hours ago is now optionally using chokidar.

https://github.com/rollup/rollup-watch/pull/39

**Comment by kzc** [comment]
> I'm pretty sure it's not caused by chokidar though,

Rollup uses a lot of memory. Some of its modules might retain state. If chokidar or your script inadvertently retain references to old rollup builds you will run out of memory and experience the crash above. That's why I suggested running each action in a its own node vm.

**Comment by jo-sm** [comment]
As far as I know, my code doesn't keep references to older Rollup builds, outside of the cache. Here is the code that I use to rollup (not including functions and variables above/outside the scope of this function):

```js
function buildJs(wss) {
  return function() {
    // On JS file changes, Rollup bundle
    log('Rolling JS up...');

    rollup.rollup({
      entry: 'assets/js/App.js',
      cache: cache,
      plugins: [
        nodeResolve({
          main: true,
          jsnext: false,
          preferBuiltins: false,
          browser: true
        }),
        commonjs(),
        json({
          include: '<removed>'
        }),
        babel(),
        replace({
          'process.env.NODE_ENV': JSON.stringify('development')
        }),
      ],
    }).then(bundle => {
      // Save bundle in cache for incremental bundling later
      // Dramatically speeds up bundle creation
      cache = bundle;

      bundle.write({
        format: 'cjs',
        moduleName: '<removed>',
        dest: 'static/js/bundle.js',
      }).then(() => {
        log('JS Bundle written');
      }).then(reload('js', wss));
    }).catch(e => {
      log(e.stack);

      const stack = e.stack.replace(/\[[0-9]{0,2}m/g, '');
      sendError(stack, wss)();
    });
  };
}
```

**Comment by jo-sm** [comment]
I'll do some more investigation into the memory leak and analyze some heaps to see if I can get a more specific cause.

**Comment by Ventajou** [comment]
I started noticing the same issue recently, I believe it may be linked to when I started using the cache for incremental builds.

**Comment by cfebs** [comment]
I'm almost positive it's the usage of cache. I did a little more research into the memory usage: https://github.com/rollup/rollup/issues/883#issuecomment-274186795

My initial guess was that `new Bundle` instances from each rollup call are not correctly cleaning up the module cache initialized here: https://github.com/rollup/rollup/blob/80da8ecf3e377820a5fc7ca6fcff8a273339614c/src/Bundle.js#L26-L31  at least that is what I want to try debugging myself soon!

**Comment by jo-sm** [comment]
I didn't comment before since I'm still trying to figure out a good way to analyze the data, but I do have per-minute heap snapshots from a single session from beginning to time of crash. Does anyone here know a good way to analyze this data? I can't easily send the data since they total 20GB of snapshot data and the Chrome debug tools for this aren't good at doing high level comparisons beyond numbers.

**Comment by cfebs** [comment]
I tried another approach with a very minimal test project: https://github.com/cfebs/rollup-cache-gc-test

It simply builds the project any number of times you want. At about 2450 iterations i would get the fatal, very consistently.

https://github.com/cfebs/rollup-cache-gc-test/blob/master/rollupnocachesubsequentheaps.tar.gz contains 2 heap dumps of 2 subsequent builds which will hopefully aid in analysis

This was using node v6.5.0 on linux.

**Comment by Rich-Harris** [comment]
This can be closed now, thanks to #1470

**Comment by Mwni** [comment]
This has not been fixed. It still crashes with the most recent version. https://github.com/cfebs/rollup-cache-gc-test still crashes after 2.5k iterations.
